agent/agent.go

```diff
@@ -488,75 +496,50 @@ func (a *Agent) StartOutput(pluginName string, config map[string]interface{}) er
 
 	output := pluginCreator()
 
-	if config != nil {
-		configJSON, err := validateStructConfig(reflect.ValueOf(output), config)
-		if err != nil {
-			return fmt.Errorf("could not update output plugin %s with error: %s", pluginName, err)
-		}
-
-		json.Unmarshal(configJSON, &output)
-	}
-
 	ro := models.NewRunningOutput(pluginName, output, &outputConfig,
 		a.Config.Agent.MetricBatchSize, a.Config.Agent.MetricBufferLimit)
 
 	ro.Init()
 
 	err := a.connectOutput(a.Context, ro)
 	if err != nil {
-		return nil
+		return err
 	}
 
-	a.RunSingleOutput(ro, a.Context)
+	err = a.RunSingleOutput(ro, a.Context)
+	if err != nil {
+		return err
+	}
 
 	// add new output to outputunit
 	a.ou.outputs = append(a.ou.outputs, ro)
 
 	return nil
 }
 
-// updateStructValuesHelper updates the reflect config, returning the original plugin otherwise
-func updateStructValuesHelper(pluginPtr reflect.Value, newConfig map[string]interface{}) (reflect.Value, map[string]interface{}, error) {
-	plugin := pluginPtr.Elem() // extract Value of type interface{} from Value pointer to interface
+// generateTomlKeysMap generates a map with the toml keys for a specific plugin
+func generateTomlKeysMap(structPtr reflect.Value, config map[string]interface{}) (map[string]interface{}, error) {
+	strct := structPtr.Elem()
 	tomlMap := map[string]interface{}{}
-	pType := plugin.Type()
-
-	if plugin.Kind() == reflect.Struct {
+	pType := strct.Type()
 
-		// initial check for errors, loop through twice to ensure that
-		// all fields are valid before updating
-		for configKey, configValue := range newConfig {
-			pluginField := plugin.FieldByName(configKey) // cast new value as Value
-			reflectedNew := reflect.ValueOf(configValue)
+	for configKey, configValue := range config {
+		field, found := pType.FieldByName(configKey)
 
-			// if any errors found, return original plugin
-			if !(pluginField.IsValid()) {
-				return pluginPtr.Elem(), map[string]interface{}{}, fmt.Errorf("invalid field name %s", configKey)
-			} else if !(pluginField.CanSet()) {
-				return pluginPtr.Elem(), map[string]interface{}{}, fmt.Errorf("unsettable field %s", configKey)
-			} else if !(pluginField.Type() == reflectedNew.Type()) {
-				return pluginPtr.Elem(), map[string]interface{}{}, fmt.Errorf("value type mismatch for field %s", configKey)
-			}
+		if !found {
+			return map[string]interface{}{}, fmt.Errorf("field %s did not exist on plugin", configKey)
```

Comment:
If you're going to return an error, not sure if this is being used, but you should probably return nil instead of an empty map. Not super important or anything though